src/opentelemetry.erl

```diff
@@ -0,0 +1,214 @@
+
+-type http_headers()       :: [{unicode:unicode_binary(), unicode:unicode_binary()}].
+
+set_default_tracer(Tracer) ->
+    persistent_term:put({?MODULE, default_tracer}, Tracer).
+
+get_tracer() ->
+    persistent_term:get({?MODULE, default_tracer}, {ot_tracer_noop, []}).
+
+get_tracer(_Name) ->
+    persistent_term:get({?MODULE, default_tracer}, {ot_tracer_noop, []}).
+
+-spec timestamp() -> integer().
+timestamp() ->
+    erlang:system_time(nanosecond).
+
+-spec links([{TraceId, SpanId, Attributes, TraceState}]) -> links() when
+      TraceId :: trace_id(),
+      SpanId :: span_id(),
+      Attributes :: attributes(),
+      TraceState :: tracestate().
+links(List) ->
+    [link(TraceId, SpanId, Attributes, TraceState) || {TraceId, SpanId, Attributes, TraceState} <- List,
+                                                      is_integer(TraceId),
+                                                      is_integer(SpanId),
+                                                      is_list(Attributes),
+                                                      is_list(TraceState)].
+
+-spec link(TraceId, SpanId, Attributes, TraceState) -> link() | undefined when
+      TraceId :: trace_id(),
+      SpanId :: span_id(),
+      Attributes :: attributes(),
+      TraceState :: tracestate().
+link(TraceId, SpanId, Attributes, TraceState) when is_integer(TraceId),
+                                                   is_integer(SpanId),
+                                                   is_list(Attributes),
+                                                   is_list(TraceState) ->
+    #link{trace_id=TraceId,
+          span_id=SpanId,
+          attributes=Attributes,
+          tracestate=TraceState};
+link(_, _, _, _) ->
+    undefined.
+
+-spec event(Name, Attributes) -> event() | undefined when
+      Name :: unicode:unicode_binary(),
+      Attributes :: attributes().
+event(Name, Attributes) when is_binary(Name),
+                             is_list(Attributes) ->
+    #event{name=Name,
+           attributes=Attributes};
+event(_, _) ->
+    undefined.
+
+-spec timed_event(TimeUnixNano, Name, Attributes) -> timed_event() | undefined when
+      TimeUnixNano :: integer(),
+      Name :: unicode:unicode_binary(),
+      Attributes :: attributes().
+timed_event(TimeUnixNano, Name, Attributes) when is_integer(TimeUnixNano),
+                                                 is_binary(Name),
+                                                 is_list(Attributes) ->
+    timed_event(TimeUnixNano, event(Name, Attributes));
+timed_event(_, _, _) ->
+    undefined.
+
+-spec timed_event(TimeUnixNano, Event) -> timed_event() | undefined when
+      TimeUnixNano :: integer(),
+      Event :: event().
+timed_event(TimeUnixNano, Event) when
+      is_integer(TimeUnixNano),
+      is_record(Event, event) ->
+    #timed_event{time_unixnano=TimeUnixNano,
+                 event=Event};
+timed_event(_, _) ->
+    undefined.
```

Comment:
So we naturally would never want to fail operations since instrumentation should never break application code, so I understand why these functions just return `undefined` if there are any input mismatches. With that said, one of the common problems w/ Spandex is that it could be difficult to test that your tracing code worked the way you expected it to without having to actually set up tracing somewhere and see if the data arrives as you expect. In the future, I think it would be nice to have a configuration that tells it to either log or fail on incorrect input to these API functions. The end users would never want this (I imagine), but it could give library authors much better piece of mind that the instrumentation code they are adding isn't going to just silently fail for their users.

Reply:
Hm, yea, I think I get what you mean. Can definitely provide some additional tooling for easing library development.